File: erpnext_custom/erpnext_custom/erpnext_custom/report/core_tax/core_tax.py
"""Daftar faktur pajak keluaran untuk impor Coretax + status kelengkapan datanya.

XML-nya dibangun di erpnext_custom.coretax; report ini cuma jendela lihat-dulu supaya
ketahuan invoice mana yang datanya belum lengkap sebelum file diupload ke Coretax.
"""

import frappe

from erpnext_custom.coretax import collect

COLUMNS = [
	{"label": "Invoice", "fieldname": "invoice", "fieldtype": "Link", "options": "Sales Invoice", "width": 160},
	{"label": "Tanggal", "fieldname": "posting_date", "fieldtype": "Date", "width": 95},
	{"label": "Pembeli", "fieldname": "customer_name", "fieldtype": "Data", "width": 200},
	{"label": "NPWP", "fieldname": "buyer_tin", "fieldtype": "Data", "width": 140},
	{"label": "NITKU", "fieldname": "buyer_nitku", "fieldtype": "Data", "width": 170},
	{"label": "Kode Trx", "fieldname": "trx_code", "fieldtype": "Data", "width": 70},
	{"label": "DPP", "fieldname": "dpp", "fieldtype": "Currency", "options": "Company:company:default_currency", "width": 120},
	{"label": "DPP Nilai Lain", "fieldname": "dpp_other", "fieldtype": "Currency", "options": "Company:company:default_currency", "width": 130},
	{"label": "PPN", "fieldname": "ppn", "fieldtype": "Currency", "options": "Company:company:default_currency", "width": 120},
	{"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 320},
]


def execute(filters=None):
	filters = filters or {}
	if not (filters.get("company") and filters.get("from_date") and filters.get("to_date")):
		return COLUMNS, []

	company, _vat_rate, entries = collect(filters)
	data = []
	for e in entries:
		si = e["si"]
		data.append(
			{
				"invoice": si.name,
				"company": company.name,
				"posting_date": si.posting_date,
				"customer_name": si.customer_name,
				"buyer_tin": si.tax_id,
				"buyer_nitku": frappe.db.get_value("Customer", si.customer, "custom_nitku"),
				"trx_code": si.get("custom_coretax_trx_code"),
				"dpp": e["dpp"],
				"dpp_other": sum(r["other"] for r in e["rows"]),
				"ppn": e["ppn"],
				"status": "Siap" if not e["missing"] else "Kurang: " + ", ".join(e["missing"]),
			}
		)
	return COLUMNS, data

File: erpnext_custom/erpnext_custom/erpnext_custom/report/core_tax/core_tax.js
// Filter + tombol unduh XML. File XML-nya dibangun di server (erpnext_custom.coretax),
// di sini cuma disimpan jadi file lewat Blob -- tidak lewat File doctype supaya file
// pajak tidak nyangkut di lampiran situs.
frappe.query_reports["Core Tax"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Perusahaan"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "from_date",
			label: __("Dari Tanggal"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.month_start(),
		},
		{
			fieldname: "to_date",
			label: __("Sampai Tanggal"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.month_end(),
		},
		{
			fieldname: "vat_rate",
			label: __("Tarif PPN Dilaporkan"),
			fieldtype: "Select",
			options: "12\n11",
			default: "12",
		},
	],

	onload(report) {
		report.page.add_inner_button(__("Download XML"), () => {
			frappe.call({
				method: "erpnext_custom.coretax.export_xml",
				args: { filters: report.get_filter_values(true) },
				freeze: true,
				freeze_message: __("Menyusun XML Coretax..."),
				callback(r) {
					if (!r.message) return;
					const link = document.createElement("a");
					link.href = URL.createObjectURL(
						new Blob([r.message.xml], { type: "application/xml" })
					);
					link.download = r.message.filename;
					link.click();
					URL.revokeObjectURL(link.href);
					frappe.show_alert({
						message: __("{0} faktur diexport, {1} dilewati", [
							r.message.ok,
							r.message.skipped,
						]),
						indicator: r.message.skipped ? "orange" : "green",
					});
				},
			});
		});
	},

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "status" && data) {
			const color = data.status === "Siap" ? "green" : "red";
			value = `<span style="color: var(--text-on-${color}); background: var(--bg-${color}); padding: 0 6px; border-radius: 4px">${value}</span>`;
		}
		return value;
	},
};
